internal/controllers/synthesis/lifecycle.go

```diff
@@ -160,34 +160,40 @@ func (c *podLifecycleController) Reconcile(ctx context.Context, req ctrl.Request
 	return ctrl.Result{}, nil
 }
 
-func (c *podLifecycleController) shouldDeletePod(logger logr.Logger, comp *apiv1.Composition, syn *apiv1.Synthesizer, pods *corev1.PodList) (logr.Logger, *corev1.Pod /* exists */, bool) {
+func shouldDeletePod(logger logr.Logger, comp *apiv1.Composition, syn *apiv1.Synthesizer, pods *corev1.PodList) (logr.Logger, *corev1.Pod, bool /* exists */) {
 	if len(pods.Items) == 0 {
 		return logger, nil, false
 	}
 
 	// Only create pods when the previous one is deleting or non-existant
+	var onePodDeleting bool
 	for _, pod := range pods.Items {
-		pod := pod
-		if pod.DeletionTimestamp != nil {
-			continue // already deleted
-		}
-
 		if comp.DeletionTimestamp != nil {
 			logger = logger.WithValues("reason", "CompositionDeleted")
 			return logger, &pod, true
 		}
 
+		// Allow a single extra pod to be created while the previous one is terminating
+		// in order to break potential deadlocks while avoiding a thundering herd of pods
+		// TODO: e2e test for this
+		pod := pod
```

Comment:
Should the shadowing still be done before the composition deletion check? Technically it's fine because we don't care which one of the pods we're returning, but just want to know what you think

Reply:
Good call 👍